### DISCUSSION TITLE
WPF Community goals for 23H1 (July 22 - Dec 22)

### DISCUSSION THREAD

**Discussion by gurpreet-wpf** [dotnet/wpf#6744] [Announcements]
Hi All,

As mentioned in our last post, we are hereby announcing the goals for community engagement towards which our efforts will be directed for the next 6 months (July 22 - Dec 22):

## Test Repository Migration

|   | Description |
| ------------- |:-------------|
| What      | WPF codebase has more than 30K integration tests. These tests ensure sanity of the build with various OS and .NET framework combination matrix. As part of open sourcing the test infrastructure, we would be moving all the tests to Github. This would also enable community to add their own tests to the test repo.     |
| Why      | Test repository migration is aimed at ensuring that community contributors add new features /fix existing issues without breaking any existing functionality.  |
| How      | WPF integration tests are divided into 2 categories - <br /> 1. DRTs: 98 tests that check the basic features of WPF controls. <br /> 2. Feature tests & PCases: 30K+ tests that cover in-depth testing for each feature <br /><br />Since there are large number of total tests for WPF, we're planning to first migrate the DRTs. This will help the community contributors to effectively test their changes against those basic tests locally and remediate anything that is breaking the tests. Once the basic testing is done, we'll move the feature tests area-by-area such as 2D, Button, Stackpanels, etc.<br /><br /> This effort is being tracked via [ this Github project](https://github.com/orgs/dotnet/projects/145). |
| When      | 23H1 (Dec22)  |
|||

----
## PR test automation

|   | Description |
| ------------- |:-------------|
| What      | DRTs need to be run on each submitted PR to ensure that the changes do not cause any regressions. Almost all major dotnet repositories do have this automation in place. We do NOT plan on running all the tests (Feature tests and p-cases) since that would take exorbitant amount of time (~2 days) to complete. The end goal of this exercise is to make sure that we have a framework in place that allows running tests on incoming PRs. Future efforts would be concentrated on reducing the test-run time and (possibly) include more unit tests in the repository as well.  |
| Why      | This would reduce the turn-around times on PRs, thereby increasing the velocity at which new changes / fixes can be merged into the repository.  |
| How      | First stage of this plan involves open sourcing the DRTs and the further steps would allow to create a pipeline that would - <br />- build the PR branch<br />- download the test binaries on the build machine<br />- run the tests<br />- update the PR with the results of the tests<br /><br /> This effort is being tracked via [ this Github project](https://github.com/orgs/dotnet/projects/144). |
| When      | 23H1 (Dec22)   |
|||

----
## Community Projects

|   | Description |
| ------------- |:-------------|
| What      | Issues and bug fixes that community would like to drive to improve product stability and usability. The effort in this front is to make sure that community efforts are aligned in the right direction and WPF keeps evolving with the needs of the community.  |
| Why      | This is how community can shape the future of WPF. Fixing issues and developing new features that impacts a huge number of developers will ensure that community needs are addressed appropriately.  |
| How      | Similar to the process we followed in the last year, we'll seek inputs from the community to choose issues/PRs so that they can be prioritized accordingly. Depending on the backlog of items, we'll aim to close the issues from last community run survey in 23H1 and then proceed for new ones.<br /><br  /> This effort is being tracked via [ this Github project](https://github.com/orgs/dotnet/projects/146). |
| When      | Ongoing activity    |
|||

----
## Support ImplicitUsings in WPF

|   | Description |
| ------------- |:-------------|
| What      | Support ImplicitUsings for WPF. This will ensure newly created WPF projects are able to utilize this feature. |
| Why      | Ensures that WPF is at par with other .NET offerings.  |
| How      | We are building an inventory of items which are required to support ImplicitUsings and will publish them on GitHub project as we progress. <br /><br  /> This effort is being tracked via [ this Github project](https://github.com/orgs/dotnet/projects/105).|
| When      | 23H1 (Dec22) |
|||

Along with these tasks, we will also invest to make sure WPF is supported on .NET 7, make accessibility improvements and continue to triage the issues and PRs in our backlog within SLAs. 
	We would also be evaluating the efforts required to support new features and enhancements for WPF such as ARM64 performance improvements, supporting .NET Upgrade Assistant and Win11 theming.
	
The progress for all these items will be tracked via GitHub projects and we will update and receive feedback through discussions and upcoming community calls.

Thanks!

Edit - Fixed formatting issues
Edit 2 - Clarified timelines

**Comment by bgrainger** [comment]
There are two links in the announcement post to https://github.com/orgs/dotnet/projects/146 but I don't have the ability to view that project. Could it be made public for the community to view?

**Reply by pchaurasia14** [reply]
@bgrainger - Yes. We're in the middle of getting the breakdown task list for each work item. 

Will be made public by early next week.

**Comment by wstaelens** [comment]
this:
https://github.com/dotnet/wpf/discussions/6542#discussioncomment-3075510

**Comment by pchaurasia14** [comment]
# UPDATE - July'22 Sprint 

As of July'22, below are the updates on our half yearly goals - 

## Test Repo Migration
|  | Description |
| ------------- |:-------------|
| What's complete      | Out of total 98 DRTs, we have successfully migrated 50 DRTs in the [public repo](https://github.com/dotnet/wpf-test).|
| What's next    | We are actively working on migrating the remaining DRTs as well. We have some dependencies on our partners which we need to close before we can migrate the remaining DRTs.     |
----

## PR Test Automation
|  | Description |
| ------------- |:-------------|
| What's complete      | - We have successfully integrated the Azure pipelines on GitHub which would enable automatic run of DRTs (whichever have been moved to public repo). The logs could be seen under `Checks -> Azure Pipelines`. <br /> - Build pipeline has also been enabled on WPF test repo that allows us to create test binaries to be used in WPF build |
| What's next    |  - As of now, DRT failures are not failing the build (so as to not block other .NET builds). Going forward, we'll be failing the pipeline run, if the tests are not successful. <br /> - We also intend to improve the reporting of the test execution, to avoid manual lookup in logs for failures.     |
----

## Community Projects
|  | Description |
| ------------- |:-------------|
| What's complete      | Out of 24 issues and PRs which we selected for this community project, 17 have been completed/closed|
| What's next    | - We're targeting remaining PR(s) for upcoming CTPs. <br /> - Followed by investigating the issues mentioned in the list.    |
----

## Support ImplicitUsings in WPF
|  | Description |
| ------------- |:-------------|
| What's complete      | We have gathered the requirements for support of ImplicitUsings |
| What's next    | Working on sketching initial implementation |

----

**Comment by kant2002** [comment]
Are there other activity from community which was accepted after that? Would be good to know what happens, since there still worry that things get sidetracked. (Maybe unreasonable)

**Reply by pchaurasia14** [reply]
Do you mean, other community issues and PR(s) (apart from the ones that were nominated for community run projects)?

If yes, then nothing more has been taken (except, no-code PRs and low-impact changes) until we get all the items in the current pipeline cleared off. 

Also. once we clear off the current pipeline, we'll start prioritizing items for the next release based on inputs from community.

**Reply by wstaelens** [reply]
still waiting for the things I've posted 🧔

**Reply by pchaurasia14** [reply]
@wstaelens  - We are reviewing XPS PRs as well. Will publish the reviews soon. Many of XPS PRs have merge conflicts (which are pending on author to be resolved) so that slows down the review process considerably.

**Reply by kant2002** [reply]
> Do you mean, other community issues and PR(s) (apart from the ones that were nominated for community run projects)?

Yes. I do mostly mean process of 
- engaging with community
- transparency.

I can understand that things maybe slow, but right now from the outside appears that nothing is happens. And because things are slow, even after month answer "we still migrating DRT" ? Okay, maybe not all DRT is ported, but accepting contributions and porting DRT can be done in parallel.

Because some DRT now run as part of PR CI I think areas which covered by them maybe require less often runs of regression tests.

**Reply by pchaurasia14** [reply]
>  maybe not all DRT is ported, but accepting contributions and porting DRT can be done in parallel.

Yes, we'll be soon accepting contributions on that front as well. Our main focus was to complete the porting before we embarked on accepting contributions so as to bring a task to closure before picking up next thing.

In fact, we are nearing the completion of majority of DRTs and will start reviewing the DRT PRs asap.

**Comment by fiza-azmi** [comment]
# UPDATE - December '22

As of December '22, below are some highlights of the progress we have made:

## Test Repo Migration
|  | Description |
| ------------- |:-------------|
| Update        | We’re thrilled to announce that all DRTs have now been migrated from internal to [ GitHub Repo](https://github.com/dotnet/wpf-test).</br>These tests can now be run locally by the community contributors to resolve issues in incoming PRs and speed up the merge pipeline.  |
| What's next    | More focus on porting Microsuite Test, solidifying the underlying infrastructure.    |
----

## PR Test Automation
|  | Description |
| ------------- |:-------------|
| Update      | We are now running these tests as a part of build on every incoming PR. </br>DRT failures are not failing the build (so as to not block other .NET builds) |
| What's next    | Going forward, the test results will be commented on respective PRs to avoid manual lookup in logs for failures.    |
----

## Community Projects
|  | Description |
| ------------- |:-------------|
| Update    | The community project is set to be a success. 24 issues and PRs were selected, 20 of which have been completed/closed. |
| What's next    | A second round of community projects is being planned. Keep an eye for an upcoming post in the Discussion Forum    |
----

## Support ImplicitUsings in WPF
|  | Description |
| ------------- |:-------------|
| Update     | WPF now supports ImplicitUsings (in .NET8 preview). All newly created WPF projects will be able to utilize this feature post release.  |
----

## Mid-year Survey 2022
|  | Description |
| ------------- |:-------------|
| Update     | We are happy to share that the community actively participated in our [Community Mid-Year Survey 2022](https://github.com/dotnet/wpf/discussions/7140). It's an incredible opportunity for us to understand ourselves, and pivot for the future. We hear you on introducing a better theming for WPF controls, support for newer controls (like WPF FolderBrowserDialog) and performance improvements. Faster responses and improving PR merge rate are also top of mind. |
----

## Upcoming Activities
1.	**Community Call**: Join the WPF team for an open house celebration of the new features in our Community Call on 15th of December 2022  
2.	**.NET 7 Servicing**: We’re aiming to address issues to improve stability of WPF on .NET7
3.	**Test suite Contributions**: We will actively be looking for additional resources to solve technical issues for a more optimized experience. As next steps to this, we would be inviting the community to make contributions to the test suite.  
4.	**Roadmap**: Watch out this space for more on the Roadmap for 23H2 and Focus Areas for PR prioritization

**Comment by NCLnclNCL** [comment]
.
